src/CTA.Rules.Models/Tokens/NodeToken.cs

```diff
@@ -52,7 +52,28 @@ public NodeToken()
         public List<ProjectLevelAction> ProjectLevelActions { get; set; }
         public List<ProjectLevelAction> ProjectFileActions { get; set; }
 
-        public NodeToken Clone() => (NodeToken)this.MemberwiseClone();
+        public NodeToken Clone()
+        {
+            NodeToken cloned = (NodeToken)this.MemberwiseClone();
+            cloned.TextChanges = cloned.TextChanges.Select(textChange => textChange.Clone()).ToList();
+            cloned.TargetCPU = cloned.TargetCPU.ToList();
+            cloned.AttributeActions = cloned.AttributeActions.Select(action => action.Clone()).ToList();
+            cloned.AttributeListActions = cloned.AttributeListActions.Select(action => action.Clone()).ToList();
+            cloned.ClassDeclarationActions = cloned.ClassDeclarationActions.Select(action => action.Clone()).ToList();
+            cloned.InterfaceDeclarationActions = cloned.InterfaceDeclarationActions.Select(action => action.Clone()).ToList();
+            cloned.MethodDeclarationActions = cloned.MethodDeclarationActions.Select(action => action.Clone()).ToList();
+            cloned.ElementAccessActions = cloned.ElementAccessActions.Select(action => action.Clone()).ToList();
+            cloned.MemberAccessActions = cloned.MemberAccessActions.Select(action => action.Clone()).ToList();
+            cloned.UsingActions = cloned.UsingActions.Select(action => action.Clone()).ToList();
+            cloned.IdentifierNameActions = cloned.IdentifierNameActions.Select(action => action.Clone()).ToList();
+            cloned.InvocationExpressionActions = cloned.InvocationExpressionActions.Select(action => action.Clone()).ToList();
+            cloned.NamespaceActions = cloned.NamespaceActions.Select(action => action.Clone()).ToList();
+            cloned.ObjectCreationExpressionActions = cloned.ObjectCreationExpressionActions.Select(action => action.Clone()).ToList();
+            cloned.PackageActions = cloned.PackageActions.Select(action => action.Clone()).ToList();
+            cloned.ProjectLevelActions = cloned.ProjectLevelActions.Select(action => action.Clone()).ToList();
+            cloned.ProjectFileActions = cloned.ProjectFileActions.Select(action => action.Clone()).ToList();
```

Comment:
What would happen if any of these actions are empty or null? Any checks we need to add here?

Reply:
That's good point. Looks like`NodeToken`'s constructor initializes all the actions to empty lists so they won't be null, unless there are places we are explicitly setting the action lists to null after `NodeToken` objects are initialized, I can add a null check to be safe. As for empty list, I think `Select` works with empty list and wont throw any exception, and it'll return an empty list, so should be safe there.